beacon_chain/utils/ssz/src/tree_hash.rs

```diff
@@ -0,0 +1,157 @@
+const CHUNKSIZE: usize = 128;
+const HASHSIZE: usize = 32;
+
+pub trait TreeHash {
+    // Note: it would be nice to have a default trait implementation here
+    // i.e. szz_encode(self) - but rust complains it does not know
+    // the size of 'self'.  Not sure if there's a way around this.
+
+    fn tree_hash(&self) -> Vec<u8>;
+}
+
+// python example:  Note - I'm seeing some inconsistencies
+// between this and the 'Tree Hash' section in the SSZ spec.
+// So, I imagine it will change.
+/* def merkle_hash(lst):
+    # Concatenate list into data
+    if len(lst[0]) != next_power_of_2(len(lst[0])):
+        lst = [extend_to_power_of_2(x) for x in lst]
+    data = b''.join(lst)
+    # Add padding
+    data += b'\x00' * (CHUNKSIZE - (len(data) % CHUNKSIZE or CHUNKSIZE))
+    assert len(data) % CHUNKSIZE == 0
+    # Store length (to compensate for non-bijectiveness of padding)
+    datalen = len(lst).to_bytes(32, 'big')
+    # Convert to chunks
+    chunkz = [data[i:i+CHUNKSIZE] for i in range(0, len(data), CHUNKSIZE)]
+    chunkz = [None] * next_power_of_2(len(chunkz)) + chunkz + [b'\x00' * CHUNKSIZE]
+    for i in range(len(chunkz)//2 - 1, 0, -1):
+        chunkz[i] = hash(chunkz[i*2] + chunkz[i*2+1])
+    return hash(chunkz[1] + datalen) */
+
+/**
+ * Returns a 32 byte hash of 'list', a vector of byte vectors.
+ * Note that this will consume 'list'.
+ * */
+pub fn merkle_hash(list: &mut Vec<Vec<u8>>) -> Vec<u8> {
+    // flatten list
+    let data = &mut list_to_blob(list);
+
+    // data should be divisible by CHUNKSIZE
+    assert_eq!(data.len() % CHUNKSIZE, 0);
```

Comment:
Panics (caused by assertion failure) in Rust terminate the current thread and are not easily "catchable" therefore we practically never use them in Lighthouse unless it's an overwhelmingly critical issue (e.g., can't write to filesystem). We prefer `Option` or `Result` instead. There's some info on the Rust philosophy [here](https://doc.rust-lang.org/nomicon/unwinding.html) and a local example [here](https://github.com/sigp/lighthouse/blob/77ac5f77bb693aa119952d033a01ebf5ec659c2a/beacon_chain/utils/ssz/src/decode.rs#L66-L88). They're completely fine in tests though, because cargo uses panics to indicate test failure.